### PR TITLE
PERF: reindex default fill_value dtype

### LINKED ISSUE #45857: PERF: reindex unnecessarily introduces block with new dtype, preventing consolidation
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this issue exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this issue exists on the main branch of pandas.

### Reproducible Example

I frequently wrap large arrays with a DataFrame, do some operations on them, and then extract the `.values` as a numpy array for e.g. passing to a `nopython` numba function or just for fast indexing (`.values[i, j]` is often much faster than `.iloc[i, j]`).

I recently got bit by an issue where Pandas unexpectedly added a new block with a different dtype to my wrapped array. This caused `.values` to slow down by a factor of 1000x because it has to build a new array every time it is called rather than simply returning the single existing block.

Specifically, if you have a single float32 block and you reindex it, Pandas adds a float64 block to the BlockManager to hold the NaNs. This seems like potentially a bad default -- in the case where there is an exisiting block the type of that existing block so long as it's possible to store the fill_value in such a block.

Example with consolidation not working:
```
df = pd.concat([
    pd.DataFrame(np.zeros((1000, 1000), dtype='f4')),
], axis=1).reindex(columns=np.arange(5, 1005))
print(df._data.nblocks) # 2

df.values
print(df._data.nblocks) # 2

%timeit df.values # 2.74ms
```

Example with consolidation working:
```
df = pd.concat([
    pd.DataFrame(np.zeros((1000, 1000), dtype='f4')),
], axis=1).reindex(columns=np.arange(5, 1005), fill_value=np.float32(np.nan))
print(df._data.nblocks) # 2

df.values
print(df._data.nblocks) # 1

%timeit df.values # 3.38 µs
```

### Installed Versions

<details>
INSTALLED VERSIONS
------------------
commit           : 945c9ed766a61c7d2c0a7cbb251b6edebf9cb7d5
python           : 3.8.12.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19042
machine          : AMD64
processor        : Intel64 Family 6 Model 85 Stepping 7, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United Kingdom.1252

pandas           : 1.3.4
numpy            : 1.21.3
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.1.1
setuptools       : 58.0.4
Cython           : 0.29.24
pytest           : 6.2.4
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.6.3
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 3.0.2
IPython          : 7.29.0
pandas_datareader: None
bs4              : 4.10.0
bottleneck       : 1.3.2
fsspec           : 2021.08.1
fastparquet      : None
gcsfs            : None
matplotlib       : 3.4.3
numexpr          : 2.7.3
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : None
pyarrow          : 3.0.0
pyxlsb           : None
s3fs             : None
scipy            : 1.6.2
sqlalchemy       : None
tables           : 3.6.1
tabulate         : None
xarray           : 0.19.0
xlrd             : 2.0.1
xlwt             : None
numba            : 0.53.1
</details>

### Prior Performance

I don't think this is a regression

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] closes #45857
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.

Default fill_value (np.nan) to match interleaved_dtype when possible to avoid upcasting and allow for block consolidation. 

Using the example from the OP:

```
import pandas as pd
import numpy as np

df = pd.concat([
    pd.DataFrame(np.zeros((1000, 1000), dtype='f4')),
], axis=1).reindex(columns=np.arange(5, 1005))
print(df._data.nblocks) 

df.values
print(df._data.nblocks)  # <- now consolidating

%timeit df.values
```

main:
```
2
2
789 µs ± 135 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops 
```

PR:
```
2
1
2.01 µs ± 48.8 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops
```

### CODE REVIEW

**Comment by lukemanley** [comment]
@jbrockmendel - any suggestions here? The test added in this PR is failing for certain builds. Unfortunately, I'm unable to reproduce the failures locally so I'm not sure where to start.

**Comment by jbrockmendel** [comment]
@lukemanley do you happen to be using a windows machine or 32bit build locally?

File: pandas/core/internals/managers.py [hunk 1]
```diff
@@ -817,6 +818,12 @@ def _make_na_block(
 
         if fill_value is None:
             fill_value = np.nan
+            dtype = interleaved_dtype([blk.dtype for blk in self.blocks])
+            if is_float_dtype(dtype):
+                # GH45857 avoid unnecessary upcasting
+                dtype = cast(np.dtype, dtype)
```
Comment (jbrockmendel) [inline pandas/core/internals/managers.py:824]: is there a risk of getting e.g. Float64Dtype here?

**Comment by jbrockmendel** [comment]
There's a comment in NDFrame._reindex_with_indexers `# TODO: speed up on homogeneous DataFrame objects (see _reindex_multi)` that i think might be useful in these cases.  Could that be helpful?

File: pandas/core/internals/managers.py (hunk 1 above)
Comment (lukemanley) [reply pandas/core/internals/managers.py:824]: I don't think there is a risk in a DataFrame of only float32 values.

**Comment by lukemanley** [comment]
> do you happen to be using a windows machine or 32bit build locally?

Nope. I'm unable to reproduce those errors on mac or 64bit windows. 

> There's a comment in NDFrame._reindex_with_indexers # TODO: speed up on homogeneous DataFrame objects (see _reindex_multi) that i think might be useful in these cases. Could that be helpful?

I'll take a look, thanks.

File: pandas/core/internals/managers.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/internals/managers.py:824]: right, im talking about a case where the DataFrame isnt all-float32, but includes at least one Float64Dtype

**Comment by lukemanley** [comment]
I’m going to close this PR. I cannot reproduce the failures locally. I tried a few different things but was unsuccessful in identifying where the float64 dtype comes from. I do not see that behavior locally.

**Pull request closed without merging**